src/util/buf.rs

```diff
@@ -11,93 +11,418 @@
+                self.start = new_cap - (cap - self.start);
+                let new_pos = self.buf.ptr().offset(self.start as isize);
+                if self.start >= cap {
+                    ptr::copy_nonoverlapping(right, new_pos, new_cap - self.start);
+                } else {
+                    ptr::copy(right, new_pos, new_cap - self.start);
+                }
+            }
+        }
+    }
+
+    /// shrink the inner buf to specified capacity.
+    pub fn shrink_to(&mut self, len: usize) {
+        assert!(len < self.capacity());
+
+        let new_cap = len + 1;
+        let to_keep = cmp::min(len, self.len());
+        unsafe {
+            if self.start <= self.end {
+                let dest = self.buf.ptr();
+                if self.start >= len {
+                    self.start = 0;
+                    self.end = to_keep;
+                    let source = self.buf.ptr().offset(self.start as isize);
+                    ptr::copy_nonoverlapping(source, dest, to_keep);
+                } else {
+                    let right_len = new_cap - self.start;
+                    if right_len >= to_keep {
+                        self.end = self.start + to_keep;
+                    } else {
+                        self.end = to_keep - right_len;
+                        let source = self.buf.ptr().offset(new_cap as isize);
+                        ptr::copy_nonoverlapping(source, dest, self.end);
+                    }
+                }
+            } else {
+                let source = self.buf.ptr().offset(self.start as isize);
+                let right_len = self.buf.cap() - self.start;
+                if right_len >= to_keep {
+                    self.start = 0;
+                    self.end = to_keep;
+                    ptr::copy(source, self.buf.ptr(), to_keep);
+                } else {
+                    self.start = new_cap - right_len;
+                    if self.end >= self.start {
+                        self.end = self.start - 1;
+                    }
+                    ptr::copy(source,
+                              self.buf.ptr().offset(self.start as isize),
+                              right_len);
+                }
+            }
+        }
+        self.buf.shrink_to_fit(new_cap);
+    }
+
+    /// Read data from `r` and fill inner buf.
+    ///
+    /// Please note that the buffer size will not change automatically,
+    /// you have to call capacity-related method to adjust it.
+    pub fn read_from<R: Read>(&mut self, r: &mut R) -> Result<usize> {
+        let mut end = self.end;
+        let mut readed;
+        {
+            let (left, right) = self.slice_append();
+            match try!(r.try_read(left)) {
+                None => return Ok(0),
+                Some(l) => readed = l,
+            }
+            end += readed;
+            if readed == left.len() && !right.is_empty() {
+                // Can't return error because r has been read into left.
```

Comment:
If there is any error with the connection, eventloop should handle it properly. If return the error here, it would break the convention that mentioned at [Read#read](https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read)
> If an error is returned then it must be guaranteed that no bytes were read.

Actually, the sys call read also follow the same convention:
>If a read() is interrupted by a signal after it has successfully read some data, it shall return the number of bytes read.